### PR TITLE
Invoking a command waits up to 30s for NuGet or another process

### LINKED ISSUE #4592: lock contention for tool project.assets.json causes tests to fail in CI

### PR DESCRIPTION
**Pull request by piotrpMSFT**
@livarcocc @jgoshi 

Fixes #4592

### CODE REVIEW

File: src/Microsoft.DotNet.Cli.Utils/FileAccessRetryer.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            if (sleepDuration == default(TimeSpan))
+            {
+                sleepDuration = TimeSpan.FromMilliseconds(10);
+            }
+            
+            while (true)
+            {
+                if (attemptsLeft < 1)
+                {
+                    throw new InvalidOperationException("Could not access assets file.");
```
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/FileAccessRetryer.cs]: Add the path in the exception?

File: src/Microsoft.DotNet.Cli.Utils/FileAccessRetryer.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                }
+
+                attemptsLeft--;
+
+                try
+                {
+                    return func();
+                }
+                catch (DirectoryNotFoundException)
```
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/FileAccessRetryer.cs]: Why do this? Why not just letting it go then?

File: test/dotnet.Tests/PackagedCommandTests.cs [hunk 3]
```diff
@@ -221,6 +221,50 @@ public void TestProjectDependencyIsNotAvailableThroughDriver()
             result.Should().Fail();        
         }
 
+        [Fact]
+        public void When_assets_file_in_use_Then_the_tool_retries_for_more_than_one_second()
```
Comment (livarcocc) [inline test/dotnet.Tests/PackagedCommandTests.cs]: How about When_assets_file_is_in_use_Then_the_tool_retries_for_more_than_one_second?

Comment (piotrpMSFT) [inline test/dotnet.Tests/PackagedCommandTests.cs]: When_assets_file_is_in_use_Then_CLI_retries_launching_the_command_for_at_least_one_second?

**Comment by piotrpMSFT** [comment]
@dotnet-bot test OSX x64 Release Build please.

**Comment by piotrpMSFT** [comment]
@dotnet-bot test Windows_NT x64 Release Build please.

**Comment by piotrpMSFT** [comment]
@dotnet-bot test OpenSUSE13.2 x64 Debug Build pleae.

File: src/Microsoft.DotNet.Cli.Utils/CommandResolution/MSBuildProject.cs [hunk 4]
```diff
@@ -124,7 +124,9 @@ public LockFile GetLockFile()
             var lockFilePath = GetLockFilePathFromProjectLockFileProperty() ??
                 GetLockFilePathFromIntermediateBaseOutputPath();
 
-            return new LockFileFormat().Read(lockFilePath);
+            return new LockFileFormat()
+                .ReadWithLock(lockFilePath)
```
Comment (krwq) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/MSBuildProject.cs]: Lock in this context can mean both ProjectLock file and ConcurrencyLock (even though ReadWithLock suggests the latter it is quite confusing on the first glance), I think a rename would make sense

Comment (piotrpMSFT) [inline src/Microsoft.DotNet.Cli.Utils/CommandResolution/MSBuildProject.cs]: This method follows the naming convention of the NuGet methods that its mimicking: https://github.com/NuGet/NuGet.Client/blob/27645cbb0be2c84dd047c8287338920bd273688f/src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs#L29

**Pull request merged**